### ISSUE TITLE
Add a way to bypass no-unused-expressions on content of expression

### ISSUE THREAD

**Issue by joovel** [eslint/eslint#9304]
**What rule do you want to change?**

`no-unused-expressions`

**Does this change cause the rule to produce more or fewer warnings?**

Ideally, this change will allow the rule to be used more frequently; the change will allow the user to configure the rule to cause fewer false-positives.

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option. I have a branch ready to go: https://github.com/joovel/eslint/tree/allow-certain-expressions

**Please provide some example code that this change will affect:**

```js
spy.should.have.been.called
spy.should.have.been.calledOnce
spy.should.have.been.calledTwice
spy.should.have.been.calledThrice
```

**What does the rule currently do for this code?**

The rule currently fails these expressions, even though they have side effects due to sinon-chai, an npm testing library.

**What will the rule do after it's changed?**

The rule will have a configuration parameter, `allowIfMatched` that permits the user to specify an array of strings. The strings are regular expressions and the rule is passed if the expression matches. For example, the code below will pass.

```js
/* eslint no-unused-expressions: ["error", {"allowIfMatches":["called(Once|Twice|Thrice)"]}] */

spy.should.have.been.called
spy.should.have.been.calledOnce
spy.should.have.been.calledTwice
spy.should.have.been.calledThrice
```

**Comment by platinumazure** [comment]
I could get behind this.

Have you considered the idea of using [esquery selectors](https://GitHub.com/estools/esquery) instead of regular expressions for configuration? That would allow users more precision. I'm not asking you to change the PR or anything, just wanted to get your thoughts.

**Comment by not-an-aardvark** [comment]
The idea of matching a regex against the source text seems a bit messy to me. For example, it will fail to catch things like

```js
foo[ /* calledOnce */ bar];
```

Personally I think it would be better to implement exceptions as a custom rule, but I'm not opposed to it if other team members are in favor.

**Comment by joovel** [comment]
@platinumazure : I hadn't considered the idea of esquery selectors. In looking at them, it seems like it would create a fairly complex selector syntax to accomplish what I want to achieve. eslint configuration has so far managed to avoid getting into the weeds of the AST, and I wanted to maintain that. It was difficult for me to grok the AST itself when I was developing the code and I don't want configuration to be as difficult. I will take some time and experiment with selectors to see if I can determine what would match the sinon-chai properties.

**Comment by joovel** [comment]
@not-an-aardvark : An embedded comment in an index sounds like an extremely unusual situation. I've looked at custom rules, but short of copy-pasting the existing `no-unused-expressions` rule, I'm not clear how to implement a custom rule that would implement the same behavior with only some limited exceptions. I'd be willing to spin up or contribute to an eslint plugin for sinon-chai users if I didn't end up copying the rest of the `no-unused-expressions` code.

**Comment by joovel** [comment]
@platinumazure : I've added a commit to the referenced branch. This changes the string format to esquery selectors instead of regular expressions. This will help address @not-an-aardvark 's concern about messy regexes.

**Comment by not-an-aardvark** [comment]
By the way, I created the [eslint-rule-composer](https://github.com/not-an-aardvark/eslint-rule-composer) package a little while ago, which would simplify the process of creating a custom rule for this without copy-pasting the `no-unused-expressions` rule.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
